### PR TITLE
feat: custom sql dimension filters

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #7564 

### Description:

Screenshot 2024-05-17 at 10 12 51

Tested:
- Add filter from sidebar
- Add filter from table header
- Add filter from table cell
- Add filter from filter select
- Add filter for unselected custom dimension ( joins necessary table )
- Add multiple filters for different types of custom dimensions

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by rephus** [comment]
`random()` doesn't work in filters, because the random it generates in `where` is different from the one in `select`  , we should probably create a cte to fix this, but I don't think is worth it. 

image

**Comment by rephus** [comment]
These custom dimension filters can't be used in dashboards, do we have plans to do it in the future ?

**Comment by rephus** [comment]
I'm using `now()` as a custom sql dimension, this returns a timestamp, but I can only filter `dates`  

image

**Comment by rephus** [comment]
I can't filter by `today` if custom sql dimension is `now()` with type `date` ,
 perhaps it is doing some timestamp formatting that not captured in the filter...

File: packages/common/src/utils/filters.mock.ts [hunk 1]
```diff
@@ -201,3 +207,28 @@ export const expectedChartWithOverrideDashboardORFilters: MetricQuery = {
...
+export const expectedFiltersWithCustomSqlDimension: Filters = {
+    dimensions: {
+        and: [
+            {
+                id: 'uuid',
+                operator: ConditionalOperator.EQUALS,
+                target: {
+                    fieldId: 'custom-sql-dimension',
+                },
+                values: [],
```
Comment (rephus) [inline packages/common/src/utils/filters.mock.ts:229]: should we add some value to test ?

Comment (rephus) [inline packages/common/src/utils/filters.mock.ts:229]: should we add some value to test ?

File: packages/common/src/utils/filters.ts [hunk 2]
```diff
@@ -102,7 +104,7 @@ export const getFilterGroupItemsPropertyName = (
 };
 
 export const getFilterTypeFromItem = (item: FilterableField): FilterType => {
-    const { type } = item;
+    const type = isCustomSqlDimension(item) ? item.dimensionType : item.type;
```
Comment (rephus) [inline packages/common/src/utils/filters.ts:107]: at this point we should probably create a `getDimensionType` method that takes a Dimension or CustomSqlDimension

**Comment by ZeRego** [comment]
> Probably something out of scope, but random() doesn't work in filters, because the random it generates in where is different from the one in select , we should probably create a cte to fix this, but I don't think is worth it.

The same limitation exists with dimensions declared in YML using custom SQL. The method `NOW()` should return the same value for all the calls in the same query execution. 

> These custom dimension filters can't be used in dashboards, do we have plans to do it in the future?

We can considerate it when users present us with good use cases. Not planned at the moment.

> I can't filter by today if custom sql dimension is now() with type date ,
perhaps it is doing some timestamp formatting that not captured in the filter...

Going to investigate this as I must have missed something.

File: packages/common/src/utils/filters.ts (hunk 2 above)
Comment (ZeRego) [reply packages/common/src/utils/filters.ts:107]: Was thinking the same, doing it after this PR

**Pull request merged**